### PR TITLE
feat(client): custom fetch for extensions

### LINKED ISSUE #17349: Deeply nested objects aren't typed when clientExtensions is enabled
### Bug description

Originally, I ran into the `RangeError: Maximum call stack size exceeded` error when trying to build my application. Using prisma version `4.9.0-integration-feat-client-ext-generic-type-api.13` as mentioned in #16600 resolved that issue. However, I am now running into an issue where deeply nested objects aren't included in my object types. 

### How to reproduce

Using the schema and code listed below, run pnpm run build. The following error is logged to the console

```
Type error: Property 'item' does not exist on type '{ id: string;  name: string; plaidItemId: string; }'.
```

### Expected behavior

_No response_

### Prisma information

```prisma
model PlaidInstitution {
  id                          String      @id
  name                        String
  logoUrl                     String?     @map("logo_url")
  items                       PlaidItem[]
  @@map("plaid_institutions")
}

model PlaidItem {
  id                          String      @id
  institutionId               String      @map("institution_id")
disabledAt                  DateTime?   @map("disabled_at")
  institution     PlaidInstitution        @relation(fields: [institutionId], references: [id], onDelete: Restrict)

  accounts                    PlaidAccount[]

  @@map("plaid_items")
  @@index([institutionId])
}

model PlaidAccount {
  id                          String      @id
  plaidItemId                 String      @map("plaid_item_id")
  name                        String
  destinations                Destination[] 
  item                        PlaidItem   @relation(fields: [plaidItemId], references: [id], onDelete: Cascade)
  @@map("plaid_accounts")
  @@index([plaidItemId])
}

model Destination {
  id                          String      @id @default(uuid())
  integration                 String
  name                        String
  accounts                    PlaidAccount[]
  
  @@map("destinations")
}
```

```ts
const destination = await db.destination.findFirst({ 
    include: { 
      accounts: { where: { item: { disabledAt: null } }, include: { item: { include: { institution: true }} }},
    }
  });
```

### Environment & setup

- OS: macOS
- Database: PlanetScale
- Node.js version: v16.17.0

### Prisma Version

```
10.1.0
```

### LINKED ISSUE #17563: Typescript issue on nested queries with clientExtensions feature
### Bug description

When clientExtensions feature is added to Prisma some nested queries start throwing Typescript errors . For example this nested query below throws an error saying "tenant" is not  available on "version" but it should be. The query works just fine without the extension and weirdly enough it works fine even when `take` and `orderBy` are omitted from the query.

### How to reproduce

```ts
  // When "take" and "orderBy" is included, typescript shows an error
  const contract = await prisma.rentContract.findUniqueOrThrow({
    where: {
      id: "",
    },
    include: {
      tenant: true,
      versions: {
        take: 1,
        orderBy: {
          createdAt: "desc",
        },
        include: {
          tenant: true,
        },
      },
    },
  });
  // Error on tenant
  contract.versions[0]?.tenant.name;

  // When "take" and "orderBy" is excluded, typescript shows correct suggestions
  const contract2 = await prisma.rentContract.findUniqueOrThrow({
    where: {
      id: "",
    },
    include: {
      tenant: true,
      versions: {
        /* take: 1,
        orderBy: {
          createdAt: "desc",
        },*/
        include: {
          tenant: true,
        },
      },
    },
  });
   // No error
  contract2.versions[0]?.tenant.name;
```

### Expected behavior

It should not throw an error

### Prisma information

```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["clientExtensions"]
}

datasource db {
  provider = "cockroachdb"
  url      = env("DATABASE_URL")
}


model Tenant {
  id          String         @id @default(cuid())
  name        String
  contracts   RentContract[]

  createdAt DateTime  @default(now())
  deletedAt DateTime?
}



model RentContract {
  id               String              @id @default(cuid())
  tenant           Tenant              @relation(fields: [tenantId], references: [id])
  tenantId         String
  versionOf        RentContract?       @relation("Versions", fields: [versionOfId], references: [id])
  versionOfId      String?
  versions         RentContract[]      @relation("Versions")

  updatedAt DateTime  @updatedAt
  createdAt DateTime  @default(now())
  deletedAt DateTime?
}

```

### Environment & setup

- OS: macOS
- Node.js version: v16.18.0

### Prisma Version

```
4.9.0
```

### PR DESCRIPTION
**Pull request by millsp**
Mainly fixes https://github.com/prisma/client-planning/issues/211
Closes https://github.com/prisma/client-planning/issues/224
Opportunistically fixes https://github.com/prisma/prisma/issues/17349
Opportunistically fixes https://github.com/prisma/prisma/issues/17563
Opportunistically fixes https://github.com/prisma/client-planning/issues/221

### CODE REVIEW

File: packages/client/scripts/default-index.d.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import runtime from '@prisma/client/runtime'
+import * as runtime from '@prisma/client/runtime'
```
Comment (millsp) [inline packages/client/scripts/default-index.d.ts:1]: Unrelated change

File: packages/client/tests/functional/_utils/setupTestSuiteEnv.ts [hunk 2]
```diff
@@ -52,9 +52,9 @@ async function copyPreprocessed(from: string, to: string, suiteConfig: Record<st
   // we adjust the relative paths to work from the generated folder
   const contents = await fs.readFile(from, 'utf8')
   const newContents = contents
-    .replace(/'..\//g, "'../../../")
-    .replace(/'.\//g, "'../../")
-    .replace(/'..\/..\/node_modules/g, "'./node_modules")
+    .replace(/'\.\.\//g, "'../../../")
```
Comment (millsp) [inline packages/client/tests/functional/_utils/setupTestSuiteEnv.ts:55]: Regex was replacing comments in tests

File: packages/client/scripts/default-index.js [hunk 3]
```diff
@@ -15,9 +15,14 @@ function defineExtension(ext) {
   return (client) => client.$extends(ext)
 }
 
+function getExtensionContext(that) {
```
Comment (millsp) [inline packages/client/scripts/default-index.js:18]: Expose `getExtensionContext` in non-generated client

File: packages/client/src/runtime/core/types/GetResult.ts [hunk 4]
```diff
@@ -50,7 +52,7 @@ export type GetFindResult<P extends Payload, A> =
             : K extends '_count'
               ? Count<GetFindResult<P, S[K]>>
               : never
-  } & (A extends { include: any } ? P['scalars'] : unknown)
+  }& (A extends { include: any } & Record<string, unknown> ? P['scalars'] : unknown)
```
Comment (millsp) [inline packages/client/src/runtime/core/types/GetResult.ts:55]: Fixes an issue with result type inference when `include` + another key is used (eg. `include` + `where` (see added test))

File: packages/engine-core/src/data-proxy/DataProxyEngine.ts [hunk 5]
```diff
@@ -210,12 +210,16 @@ export class DataProxyEngine extends Engine<DataProxyTxInfoPayload> {
           headers['X-transaction-id'] = interactiveTransaction.id
         }
 
-        const response = await request(url, {
-          method: 'POST',
-          headers: { ...headers, ...this.headers },
-          body: JSON.stringify(body),
-          clientVersion: this.clientVersion,
-        })
+        const response = await request(
```
Comment (millsp) [inline packages/engine-core/src/data-proxy/DataProxyEngine.ts:213]: Maybe it could have been simpler like `customFetch(request)(url, ...)` instead of passing a param

File: packages/client/src/runtime/RequestHandler.ts [hunk 6]
```diff
@@ -40,13 +41,13 @@ export type RequestParams = {
...
   rejectOnNotFound?: RejectOnNotFound
   transaction?: PrismaPromiseTransaction
-  customDataProxyHeaders?: Record<string, string>
   extensions: MergedExtensionsList
   args?: any
   headers?: Record<string, string>
   unpacker?: Unpacker
   otelParentCtx?: Context
   otelChildCtx?: Context
+  customFetch?: (fetch: Fetch) => Fetch
```
Comment (SevInf) [inline packages/client/src/runtime/RequestHandler.ts:50]: Could we remove `customDataProxyHeaders` straight away?

Comment (SevInf) [inline packages/client/src/runtime/RequestHandler.ts:50]: Could we maybe name it `customDataProxyFetch` or something similar, to indicate that it won't be used for binary engine?

**Review by SevInf — APPROVED** [review]

File: packages/client/tests/functional/extensions/enabled/query-accelerate.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+      },
+    })
+  })
+
+  beforeEach(() => {
+    mockedRequest = jest.fn()
+    https.request = (...args: any[]) => {
+      mockedRequest(args[0], args[1], args[2])
+      return originalRequest(args[0], args[1], args[2])
+    }
```
Comment (SevInf) [inline packages/client/tests/functional/extensions/enabled/query-accelerate.ts:35]:
```suggestion
   mockedRequest = jest.spyOn(https, 'request')
```
